{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>GeneID</th>\n",
       "      <th>100287102</th>\n",
       "      <th>653635</th>\n",
       "      <th>102466751</th>\n",
       "      <th>107985730</th>\n",
       "      <th>100302278</th>\n",
       "      <th>645520</th>\n",
       "      <th>79501</th>\n",
       "      <th>100996442</th>\n",
       "      <th>729737</th>\n",
       "      <th>102725121</th>\n",
       "      <th>...</th>\n",
       "      <th>4575</th>\n",
       "      <th>4568</th>\n",
       "      <th>4540</th>\n",
       "      <th>4541</th>\n",
       "      <th>4556</th>\n",
       "      <th>4519</th>\n",
       "      <th>4576</th>\n",
       "      <th>4571</th>\n",
       "      <th>Edited (1) or Unedited (0)</th>\n",
       "      <th>editing mechanism</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>GSM6745599</th>\n",
       "      <td>0.212625</td>\n",
       "      <td>-0.318248</td>\n",
       "      <td>-0.042237</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.117663</td>\n",
       "      <td>0.140983</td>\n",
       "      <td>-0.006097</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.267632</td>\n",
       "      <td>0.466206</td>\n",
       "      <td>-0.318275</td>\n",
       "      <td>-0.133587</td>\n",
       "      <td>0.136065</td>\n",
       "      <td>-0.604440</td>\n",
       "      <td>-0.270274</td>\n",
       "      <td>-0.375282</td>\n",
       "      <td>1</td>\n",
       "      <td>BE4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745600</th>\n",
       "      <td>-1.311857</td>\n",
       "      <td>-1.500716</td>\n",
       "      <td>-1.647234</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-1.159822</td>\n",
       "      <td>-0.267197</td>\n",
       "      <td>-1.511998</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007233</td>\n",
       "      <td>1.309818</td>\n",
       "      <td>-0.981598</td>\n",
       "      <td>-0.819478</td>\n",
       "      <td>-0.337051</td>\n",
       "      <td>-1.184536</td>\n",
       "      <td>-1.084285</td>\n",
       "      <td>-0.450273</td>\n",
       "      <td>1</td>\n",
       "      <td>BE4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745601</th>\n",
       "      <td>1.127315</td>\n",
       "      <td>1.334520</td>\n",
       "      <td>0.492762</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>1.760144</td>\n",
       "      <td>1.722682</td>\n",
       "      <td>0.688935</td>\n",
       "      <td>...</td>\n",
       "      <td>0.969260</td>\n",
       "      <td>2.153430</td>\n",
       "      <td>1.484606</td>\n",
       "      <td>1.848533</td>\n",
       "      <td>1.741280</td>\n",
       "      <td>0.865277</td>\n",
       "      <td>1.357749</td>\n",
       "      <td>1.874436</td>\n",
       "      <td>1</td>\n",
       "      <td>BE4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745602</th>\n",
       "      <td>0.365073</td>\n",
       "      <td>-0.318248</td>\n",
       "      <td>-0.577236</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>1.121401</td>\n",
       "      <td>0.498141</td>\n",
       "      <td>0.225580</td>\n",
       "      <td>...</td>\n",
       "      <td>0.282098</td>\n",
       "      <td>2.715838</td>\n",
       "      <td>0.638045</td>\n",
       "      <td>1.031012</td>\n",
       "      <td>0.913327</td>\n",
       "      <td>0.141225</td>\n",
       "      <td>0.041475</td>\n",
       "      <td>0.237141</td>\n",
       "      <td>1</td>\n",
       "      <td>ABE8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745603</th>\n",
       "      <td>0.669970</td>\n",
       "      <td>0.111740</td>\n",
       "      <td>-0.042237</td>\n",
       "      <td>1.188747</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.391410</td>\n",
       "      <td>-0.165152</td>\n",
       "      <td>0.804773</td>\n",
       "      <td>...</td>\n",
       "      <td>0.831828</td>\n",
       "      <td>1.028614</td>\n",
       "      <td>0.434101</td>\n",
       "      <td>0.974346</td>\n",
       "      <td>1.293509</td>\n",
       "      <td>0.187012</td>\n",
       "      <td>0.110753</td>\n",
       "      <td>0.624593</td>\n",
       "      <td>1</td>\n",
       "      <td>ABE8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745604</th>\n",
       "      <td>-0.092271</td>\n",
       "      <td>0.622352</td>\n",
       "      <td>1.562760</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.756406</td>\n",
       "      <td>0.804276</td>\n",
       "      <td>-0.121935</td>\n",
       "      <td>...</td>\n",
       "      <td>1.381557</td>\n",
       "      <td>0.185003</td>\n",
       "      <td>1.132972</td>\n",
       "      <td>1.712771</td>\n",
       "      <td>2.053874</td>\n",
       "      <td>0.594337</td>\n",
       "      <td>1.115277</td>\n",
       "      <td>1.137029</td>\n",
       "      <td>1</td>\n",
       "      <td>ABE8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745605</th>\n",
       "      <td>0.669970</td>\n",
       "      <td>0.622352</td>\n",
       "      <td>-1.112235</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.026414</td>\n",
       "      <td>-0.012084</td>\n",
       "      <td>0.225580</td>\n",
       "      <td>...</td>\n",
       "      <td>2.206152</td>\n",
       "      <td>0.466206</td>\n",
       "      <td>-0.349242</td>\n",
       "      <td>0.224115</td>\n",
       "      <td>0.786600</td>\n",
       "      <td>-0.666955</td>\n",
       "      <td>-0.339551</td>\n",
       "      <td>0.862063</td>\n",
       "      <td>1</td>\n",
       "      <td>Cas9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745606</th>\n",
       "      <td>-0.702064</td>\n",
       "      <td>-1.231973</td>\n",
       "      <td>-0.577236</td>\n",
       "      <td>-1.822745</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>3.41565</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-1.159822</td>\n",
       "      <td>-0.828445</td>\n",
       "      <td>-1.048644</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.817361</td>\n",
       "      <td>-0.096201</td>\n",
       "      <td>-0.984529</td>\n",
       "      <td>-0.787604</td>\n",
       "      <td>-0.413087</td>\n",
       "      <td>-1.281728</td>\n",
       "      <td>-0.997688</td>\n",
       "      <td>-0.900216</td>\n",
       "      <td>1</td>\n",
       "      <td>Cas9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745607</th>\n",
       "      <td>0.517522</td>\n",
       "      <td>0.138615</td>\n",
       "      <td>1.562760</td>\n",
       "      <td>1.188747</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.208912</td>\n",
       "      <td>-0.012084</td>\n",
       "      <td>0.457258</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.130199</td>\n",
       "      <td>0.185003</td>\n",
       "      <td>0.108853</td>\n",
       "      <td>0.397654</td>\n",
       "      <td>0.474005</td>\n",
       "      <td>-0.268787</td>\n",
       "      <td>-0.443467</td>\n",
       "      <td>-0.125313</td>\n",
       "      <td>1</td>\n",
       "      <td>Cas9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745609</th>\n",
       "      <td>-0.092271</td>\n",
       "      <td>-0.801985</td>\n",
       "      <td>-1.112235</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.886075</td>\n",
       "      <td>-1.185602</td>\n",
       "      <td>0.109742</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.367091</td>\n",
       "      <td>-0.658609</td>\n",
       "      <td>-0.198254</td>\n",
       "      <td>-0.324244</td>\n",
       "      <td>-0.945342</td>\n",
       "      <td>-0.949860</td>\n",
       "      <td>-1.188201</td>\n",
       "      <td>-0.950210</td>\n",
       "      <td>1</td>\n",
       "      <td>Utelectro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745610</th>\n",
       "      <td>-0.244720</td>\n",
       "      <td>0.582040</td>\n",
       "      <td>-0.577236</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.117663</td>\n",
       "      <td>-0.267197</td>\n",
       "      <td>0.109742</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.542496</td>\n",
       "      <td>-0.658609</td>\n",
       "      <td>0.579042</td>\n",
       "      <td>0.754766</td>\n",
       "      <td>0.575387</td>\n",
       "      <td>-0.268543</td>\n",
       "      <td>-0.339551</td>\n",
       "      <td>-0.400279</td>\n",
       "      <td>1</td>\n",
       "      <td>Utelectro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745611</th>\n",
       "      <td>-1.159409</td>\n",
       "      <td>-0.936356</td>\n",
       "      <td>-0.042237</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.117663</td>\n",
       "      <td>-0.216174</td>\n",
       "      <td>-0.816967</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.405064</td>\n",
       "      <td>-0.658609</td>\n",
       "      <td>-0.861943</td>\n",
       "      <td>-0.637086</td>\n",
       "      <td>-0.522918</td>\n",
       "      <td>-0.602974</td>\n",
       "      <td>-0.374190</td>\n",
       "      <td>-0.862721</td>\n",
       "      <td>1</td>\n",
       "      <td>BE4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745612</th>\n",
       "      <td>-0.244720</td>\n",
       "      <td>-0.304811</td>\n",
       "      <td>-0.577236</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.156084</td>\n",
       "      <td>1.161434</td>\n",
       "      <td>-0.932805</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.542496</td>\n",
       "      <td>0.185003</td>\n",
       "      <td>-0.479158</td>\n",
       "      <td>-0.164281</td>\n",
       "      <td>0.017786</td>\n",
       "      <td>-0.264147</td>\n",
       "      <td>-0.304912</td>\n",
       "      <td>-0.637749</td>\n",
       "      <td>1</td>\n",
       "      <td>BE4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745613</th>\n",
       "      <td>0.822418</td>\n",
       "      <td>1.818257</td>\n",
       "      <td>0.492762</td>\n",
       "      <td>2.192578</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>2.033890</td>\n",
       "      <td>0.804276</td>\n",
       "      <td>2.078998</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.130199</td>\n",
       "      <td>-0.939813</td>\n",
       "      <td>-0.401282</td>\n",
       "      <td>-0.694342</td>\n",
       "      <td>-1.156555</td>\n",
       "      <td>-0.068665</td>\n",
       "      <td>-0.149038</td>\n",
       "      <td>0.124655</td>\n",
       "      <td>1</td>\n",
       "      <td>BE4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745614</th>\n",
       "      <td>-1.006961</td>\n",
       "      <td>-0.439182</td>\n",
       "      <td>-1.112235</td>\n",
       "      <td>1.188747</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.026414</td>\n",
       "      <td>-0.522310</td>\n",
       "      <td>-0.585290</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.267632</td>\n",
       "      <td>-0.658609</td>\n",
       "      <td>-0.542559</td>\n",
       "      <td>-0.396846</td>\n",
       "      <td>-0.092044</td>\n",
       "      <td>-0.319092</td>\n",
       "      <td>-0.478106</td>\n",
       "      <td>-0.362783</td>\n",
       "      <td>1</td>\n",
       "      <td>ABE8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745615</th>\n",
       "      <td>2.194452</td>\n",
       "      <td>2.839480</td>\n",
       "      <td>-0.577236</td>\n",
       "      <td>2.192578</td>\n",
       "      <td>2.900209</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>4.242641</td>\n",
       "      <td>2.672633</td>\n",
       "      <td>3.508470</td>\n",
       "      <td>2.194836</td>\n",
       "      <td>...</td>\n",
       "      <td>1.518990</td>\n",
       "      <td>-0.096201</td>\n",
       "      <td>2.198502</td>\n",
       "      <td>1.854436</td>\n",
       "      <td>1.741280</td>\n",
       "      <td>2.608864</td>\n",
       "      <td>1.929288</td>\n",
       "      <td>1.624467</td>\n",
       "      <td>1</td>\n",
       "      <td>ABE8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745616</th>\n",
       "      <td>-0.092271</td>\n",
       "      <td>0.017680</td>\n",
       "      <td>-0.577236</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>4.242641</td>\n",
       "      <td>0.391410</td>\n",
       "      <td>0.345073</td>\n",
       "      <td>0.225580</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.267632</td>\n",
       "      <td>-0.096201</td>\n",
       "      <td>0.107754</td>\n",
       "      <td>-0.046818</td>\n",
       "      <td>-0.235669</td>\n",
       "      <td>0.433776</td>\n",
       "      <td>0.180030</td>\n",
       "      <td>0.249640</td>\n",
       "      <td>1</td>\n",
       "      <td>ABE8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745617</th>\n",
       "      <td>-0.854513</td>\n",
       "      <td>0.138615</td>\n",
       "      <td>0.492762</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.117663</td>\n",
       "      <td>0.396096</td>\n",
       "      <td>-0.932805</td>\n",
       "      <td>...</td>\n",
       "      <td>0.007233</td>\n",
       "      <td>-1.502221</td>\n",
       "      <td>-0.400183</td>\n",
       "      <td>-0.797048</td>\n",
       "      <td>-0.928445</td>\n",
       "      <td>0.090553</td>\n",
       "      <td>0.439821</td>\n",
       "      <td>-0.162809</td>\n",
       "      <td>1</td>\n",
       "      <td>Cas9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745618</th>\n",
       "      <td>-1.006961</td>\n",
       "      <td>0.850783</td>\n",
       "      <td>1.562760</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.208912</td>\n",
       "      <td>0.855299</td>\n",
       "      <td>-0.932805</td>\n",
       "      <td>...</td>\n",
       "      <td>0.556963</td>\n",
       "      <td>1.309818</td>\n",
       "      <td>0.920782</td>\n",
       "      <td>1.065838</td>\n",
       "      <td>1.031606</td>\n",
       "      <td>1.194703</td>\n",
       "      <td>0.872806</td>\n",
       "      <td>0.537104</td>\n",
       "      <td>1</td>\n",
       "      <td>Cas9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745619</th>\n",
       "      <td>-1.769202</td>\n",
       "      <td>-1.850082</td>\n",
       "      <td>-1.112235</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-1.889813</td>\n",
       "      <td>-1.032535</td>\n",
       "      <td>-1.859514</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.229658</td>\n",
       "      <td>-1.221017</td>\n",
       "      <td>-1.899621</td>\n",
       "      <td>-1.863662</td>\n",
       "      <td>-1.587428</td>\n",
       "      <td>-1.646929</td>\n",
       "      <td>-1.569228</td>\n",
       "      <td>-1.712614</td>\n",
       "      <td>1</td>\n",
       "      <td>Cas9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745620</th>\n",
       "      <td>0.060177</td>\n",
       "      <td>-0.439182</td>\n",
       "      <td>-1.112235</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.338582</td>\n",
       "      <td>-0.114129</td>\n",
       "      <td>0.341419</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.954794</td>\n",
       "      <td>-0.658609</td>\n",
       "      <td>-0.717918</td>\n",
       "      <td>-0.702015</td>\n",
       "      <td>-0.506021</td>\n",
       "      <td>-0.409690</td>\n",
       "      <td>-0.582023</td>\n",
       "      <td>-0.850222</td>\n",
       "      <td>1</td>\n",
       "      <td>Utelectro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745621</th>\n",
       "      <td>-0.549616</td>\n",
       "      <td>-0.506368</td>\n",
       "      <td>-1.112235</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.521079</td>\n",
       "      <td>-0.522310</td>\n",
       "      <td>-0.701128</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.954794</td>\n",
       "      <td>0.185003</td>\n",
       "      <td>-0.593133</td>\n",
       "      <td>-0.405701</td>\n",
       "      <td>-0.320154</td>\n",
       "      <td>-0.369275</td>\n",
       "      <td>-0.495426</td>\n",
       "      <td>-0.662746</td>\n",
       "      <td>1</td>\n",
       "      <td>Utelectro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745622</th>\n",
       "      <td>0.974866</td>\n",
       "      <td>0.944843</td>\n",
       "      <td>-0.577236</td>\n",
       "      <td>1.188747</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.391410</td>\n",
       "      <td>0.447118</td>\n",
       "      <td>0.457258</td>\n",
       "      <td>...</td>\n",
       "      <td>0.144666</td>\n",
       "      <td>-0.377405</td>\n",
       "      <td>-0.089227</td>\n",
       "      <td>-0.238064</td>\n",
       "      <td>-0.320154</td>\n",
       "      <td>0.131579</td>\n",
       "      <td>0.024156</td>\n",
       "      <td>-0.450273</td>\n",
       "      <td>1</td>\n",
       "      <td>Utelectro</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745623</th>\n",
       "      <td>-0.092271</td>\n",
       "      <td>0.004243</td>\n",
       "      <td>2.097759</td>\n",
       "      <td>1.188747</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>3.41565</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.703577</td>\n",
       "      <td>1.161434</td>\n",
       "      <td>0.225580</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.542496</td>\n",
       "      <td>-0.377405</td>\n",
       "      <td>0.290442</td>\n",
       "      <td>0.295538</td>\n",
       "      <td>0.203653</td>\n",
       "      <td>0.197757</td>\n",
       "      <td>-0.114399</td>\n",
       "      <td>-0.300291</td>\n",
       "      <td>1</td>\n",
       "      <td>dCas9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745624</th>\n",
       "      <td>-0.549616</td>\n",
       "      <td>-0.210751</td>\n",
       "      <td>0.492762</td>\n",
       "      <td>2.192578</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.612328</td>\n",
       "      <td>-0.216174</td>\n",
       "      <td>-0.237774</td>\n",
       "      <td>...</td>\n",
       "      <td>0.282098</td>\n",
       "      <td>-0.096201</td>\n",
       "      <td>0.569148</td>\n",
       "      <td>0.543450</td>\n",
       "      <td>0.693666</td>\n",
       "      <td>0.771382</td>\n",
       "      <td>0.803528</td>\n",
       "      <td>0.462113</td>\n",
       "      <td>1</td>\n",
       "      <td>dCas9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745625</th>\n",
       "      <td>-0.092271</td>\n",
       "      <td>-0.519805</td>\n",
       "      <td>1.027761</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.977324</td>\n",
       "      <td>-0.267197</td>\n",
       "      <td>-0.469451</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.542496</td>\n",
       "      <td>-0.658609</td>\n",
       "      <td>-0.102054</td>\n",
       "      <td>-0.185531</td>\n",
       "      <td>-0.269463</td>\n",
       "      <td>0.247574</td>\n",
       "      <td>0.110753</td>\n",
       "      <td>0.162151</td>\n",
       "      <td>1</td>\n",
       "      <td>dCas9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745626</th>\n",
       "      <td>1.432211</td>\n",
       "      <td>2.006377</td>\n",
       "      <td>2.097759</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.482659</td>\n",
       "      <td>1.212456</td>\n",
       "      <td>1.731482</td>\n",
       "      <td>...</td>\n",
       "      <td>1.381557</td>\n",
       "      <td>0.466206</td>\n",
       "      <td>1.629181</td>\n",
       "      <td>1.468400</td>\n",
       "      <td>1.563861</td>\n",
       "      <td>1.587498</td>\n",
       "      <td>1.808053</td>\n",
       "      <td>1.699458</td>\n",
       "      <td>1</td>\n",
       "      <td>BE4alone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745627</th>\n",
       "      <td>-2.378995</td>\n",
       "      <td>-2.239759</td>\n",
       "      <td>-1.647234</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-1.798564</td>\n",
       "      <td>-2.359120</td>\n",
       "      <td>-1.975353</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.916820</td>\n",
       "      <td>-1.502221</td>\n",
       "      <td>-2.095136</td>\n",
       "      <td>-1.896718</td>\n",
       "      <td>-1.511392</td>\n",
       "      <td>-1.923485</td>\n",
       "      <td>-1.655824</td>\n",
       "      <td>-1.825100</td>\n",
       "      <td>1</td>\n",
       "      <td>BE4alone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745628</th>\n",
       "      <td>0.365073</td>\n",
       "      <td>0.729849</td>\n",
       "      <td>-0.577236</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.886075</td>\n",
       "      <td>-0.114129</td>\n",
       "      <td>0.109742</td>\n",
       "      <td>...</td>\n",
       "      <td>1.656422</td>\n",
       "      <td>-0.377405</td>\n",
       "      <td>-0.251393</td>\n",
       "      <td>-1.024301</td>\n",
       "      <td>-1.384664</td>\n",
       "      <td>0.168819</td>\n",
       "      <td>0.526418</td>\n",
       "      <td>0.762076</td>\n",
       "      <td>1</td>\n",
       "      <td>BE4alone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745629</th>\n",
       "      <td>-0.549616</td>\n",
       "      <td>-0.277937</td>\n",
       "      <td>-0.042237</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.391410</td>\n",
       "      <td>0.345073</td>\n",
       "      <td>-0.353613</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.817361</td>\n",
       "      <td>-0.658609</td>\n",
       "      <td>-0.627032</td>\n",
       "      <td>-0.993607</td>\n",
       "      <td>-1.190349</td>\n",
       "      <td>-0.302731</td>\n",
       "      <td>-0.616661</td>\n",
       "      <td>-0.862721</td>\n",
       "      <td>1</td>\n",
       "      <td>ABE8alone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745630</th>\n",
       "      <td>-1.006961</td>\n",
       "      <td>-1.003542</td>\n",
       "      <td>-0.042237</td>\n",
       "      <td>-1.822745</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.794826</td>\n",
       "      <td>-1.593783</td>\n",
       "      <td>-0.932805</td>\n",
       "      <td>...</td>\n",
       "      <td>-1.229658</td>\n",
       "      <td>-0.939813</td>\n",
       "      <td>-1.498146</td>\n",
       "      <td>-1.513043</td>\n",
       "      <td>-1.317076</td>\n",
       "      <td>-1.254255</td>\n",
       "      <td>-1.240159</td>\n",
       "      <td>-1.350160</td>\n",
       "      <td>1</td>\n",
       "      <td>ABE8alone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745631</th>\n",
       "      <td>-0.549616</td>\n",
       "      <td>0.595477</td>\n",
       "      <td>1.027761</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>0.665157</td>\n",
       "      <td>-0.165152</td>\n",
       "      <td>-0.006097</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.542496</td>\n",
       "      <td>-0.658609</td>\n",
       "      <td>0.576111</td>\n",
       "      <td>0.642025</td>\n",
       "      <td>0.685218</td>\n",
       "      <td>0.578953</td>\n",
       "      <td>0.509099</td>\n",
       "      <td>0.262138</td>\n",
       "      <td>1</td>\n",
       "      <td>ABE8alone</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745632</th>\n",
       "      <td>0.822418</td>\n",
       "      <td>0.152052</td>\n",
       "      <td>0.492762</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>2.216388</td>\n",
       "      <td>-1.134580</td>\n",
       "      <td>1.615643</td>\n",
       "      <td>...</td>\n",
       "      <td>0.282098</td>\n",
       "      <td>0.185003</td>\n",
       "      <td>1.386757</td>\n",
       "      <td>1.317292</td>\n",
       "      <td>1.369546</td>\n",
       "      <td>0.938049</td>\n",
       "      <td>1.201874</td>\n",
       "      <td>1.474486</td>\n",
       "      <td>0</td>\n",
       "      <td>UT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745633</th>\n",
       "      <td>1.737108</td>\n",
       "      <td>0.541729</td>\n",
       "      <td>-0.577236</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.064835</td>\n",
       "      <td>0.294051</td>\n",
       "      <td>1.615643</td>\n",
       "      <td>...</td>\n",
       "      <td>2.618449</td>\n",
       "      <td>2.434634</td>\n",
       "      <td>2.424618</td>\n",
       "      <td>1.288369</td>\n",
       "      <td>-0.218772</td>\n",
       "      <td>3.018998</td>\n",
       "      <td>3.245562</td>\n",
       "      <td>2.811819</td>\n",
       "      <td>0</td>\n",
       "      <td>UT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745634</th>\n",
       "      <td>-1.006961</td>\n",
       "      <td>-0.748236</td>\n",
       "      <td>1.027761</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>3.41565</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.338582</td>\n",
       "      <td>-0.471287</td>\n",
       "      <td>-1.048644</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.817361</td>\n",
       "      <td>-1.502221</td>\n",
       "      <td>-0.806605</td>\n",
       "      <td>-1.298776</td>\n",
       "      <td>-1.604325</td>\n",
       "      <td>-0.500777</td>\n",
       "      <td>-0.720578</td>\n",
       "      <td>-0.537762</td>\n",
       "      <td>0</td>\n",
       "      <td>UT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745635</th>\n",
       "      <td>0.517522</td>\n",
       "      <td>-0.425745</td>\n",
       "      <td>-0.577236</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>1.063410</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.886075</td>\n",
       "      <td>-0.879467</td>\n",
       "      <td>-0.237774</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.405064</td>\n",
       "      <td>-0.096201</td>\n",
       "      <td>-0.742472</td>\n",
       "      <td>-0.869061</td>\n",
       "      <td>-0.801718</td>\n",
       "      <td>-0.835331</td>\n",
       "      <td>-0.893771</td>\n",
       "      <td>-1.000204</td>\n",
       "      <td>0</td>\n",
       "      <td>UT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745636</th>\n",
       "      <td>1.737108</td>\n",
       "      <td>0.138615</td>\n",
       "      <td>0.492762</td>\n",
       "      <td>0.184916</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.338582</td>\n",
       "      <td>-0.369242</td>\n",
       "      <td>1.152289</td>\n",
       "      <td>...</td>\n",
       "      <td>0.831828</td>\n",
       "      <td>0.466206</td>\n",
       "      <td>0.280914</td>\n",
       "      <td>0.466125</td>\n",
       "      <td>0.490902</td>\n",
       "      <td>0.140248</td>\n",
       "      <td>-0.374190</td>\n",
       "      <td>-0.150310</td>\n",
       "      <td>0</td>\n",
       "      <td>UT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GSM6745637</th>\n",
       "      <td>1.127315</td>\n",
       "      <td>-0.116691</td>\n",
       "      <td>1.027761</td>\n",
       "      <td>-0.818915</td>\n",
       "      <td>-0.773389</td>\n",
       "      <td>-0.29277</td>\n",
       "      <td>-0.235702</td>\n",
       "      <td>-0.703577</td>\n",
       "      <td>-0.981512</td>\n",
       "      <td>0.341419</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.267632</td>\n",
       "      <td>0.747410</td>\n",
       "      <td>0.177934</td>\n",
       "      <td>-0.092858</td>\n",
       "      <td>-0.108941</td>\n",
       "      <td>-0.174404</td>\n",
       "      <td>-0.045122</td>\n",
       "      <td>0.024668</td>\n",
       "      <td>0</td>\n",
       "      <td>UT</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>38 rows Ã— 39378 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "GeneID      100287102    653635  102466751  107985730  100302278   645520  \\\n",
       "GSM6745599   0.212625 -0.318248  -0.042237   0.184916   1.063410 -0.29277   \n",
       "GSM6745600  -1.311857 -1.500716  -1.647234  -0.818915   1.063410 -0.29277   \n",
       "GSM6745601   1.127315  1.334520   0.492762  -0.818915  -0.773389 -0.29277   \n",
       "GSM6745602   0.365073 -0.318248  -0.577236   0.184916  -0.773389 -0.29277   \n",
       "GSM6745603   0.669970  0.111740  -0.042237   1.188747  -0.773389 -0.29277   \n",
       "GSM6745604  -0.092271  0.622352   1.562760   0.184916  -0.773389 -0.29277   \n",
       "GSM6745605   0.669970  0.622352  -1.112235  -0.818915  -0.773389 -0.29277   \n",
       "GSM6745606  -0.702064 -1.231973  -0.577236  -1.822745   1.063410  3.41565   \n",
       "GSM6745607   0.517522  0.138615   1.562760   1.188747  -0.773389 -0.29277   \n",
       "GSM6745609  -0.092271 -0.801985  -1.112235  -0.818915  -0.773389 -0.29277   \n",
       "GSM6745610  -0.244720  0.582040  -0.577236   0.184916  -0.773389 -0.29277   \n",
       "GSM6745611  -1.159409 -0.936356  -0.042237   0.184916  -0.773389 -0.29277   \n",
       "GSM6745612  -0.244720 -0.304811  -0.577236  -0.818915   1.063410 -0.29277   \n",
       "GSM6745613   0.822418  1.818257   0.492762   2.192578   1.063410 -0.29277   \n",
       "GSM6745614  -1.006961 -0.439182  -1.112235   1.188747  -0.773389 -0.29277   \n",
       "GSM6745615   2.194452  2.839480  -0.577236   2.192578   2.900209 -0.29277   \n",
       "GSM6745616  -0.092271  0.017680  -0.577236  -0.818915  -0.773389 -0.29277   \n",
       "GSM6745617  -0.854513  0.138615   0.492762   0.184916   1.063410 -0.29277   \n",
       "GSM6745618  -1.006961  0.850783   1.562760   0.184916  -0.773389 -0.29277   \n",
       "GSM6745619  -1.769202 -1.850082  -1.112235  -0.818915   1.063410 -0.29277   \n",
       "GSM6745620   0.060177 -0.439182  -1.112235   0.184916  -0.773389 -0.29277   \n",
       "GSM6745621  -0.549616 -0.506368  -1.112235   0.184916  -0.773389 -0.29277   \n",
       "GSM6745622   0.974866  0.944843  -0.577236   1.188747   1.063410 -0.29277   \n",
       "GSM6745623  -0.092271  0.004243   2.097759   1.188747   1.063410  3.41565   \n",
       "GSM6745624  -0.549616 -0.210751   0.492762   2.192578  -0.773389 -0.29277   \n",
       "GSM6745625  -0.092271 -0.519805   1.027761  -0.818915  -0.773389 -0.29277   \n",
       "GSM6745626   1.432211  2.006377   2.097759   0.184916   1.063410 -0.29277   \n",
       "GSM6745627  -2.378995 -2.239759  -1.647234  -0.818915  -0.773389 -0.29277   \n",
       "GSM6745628   0.365073  0.729849  -0.577236   0.184916   1.063410 -0.29277   \n",
       "GSM6745629  -0.549616 -0.277937  -0.042237   0.184916   1.063410 -0.29277   \n",
       "GSM6745630  -1.006961 -1.003542  -0.042237  -1.822745  -0.773389 -0.29277   \n",
       "GSM6745631  -0.549616  0.595477   1.027761   0.184916   1.063410 -0.29277   \n",
       "GSM6745632   0.822418  0.152052   0.492762  -0.818915  -0.773389 -0.29277   \n",
       "GSM6745633   1.737108  0.541729  -0.577236  -0.818915  -0.773389 -0.29277   \n",
       "GSM6745634  -1.006961 -0.748236   1.027761  -0.818915  -0.773389  3.41565   \n",
       "GSM6745635   0.517522 -0.425745  -0.577236  -0.818915   1.063410 -0.29277   \n",
       "GSM6745636   1.737108  0.138615   0.492762   0.184916  -0.773389 -0.29277   \n",
       "GSM6745637   1.127315 -0.116691   1.027761  -0.818915  -0.773389 -0.29277   \n",
       "\n",
       "GeneID         79501  100996442    729737  102725121  ...      4575      4568  \\\n",
       "GSM6745599 -0.235702   0.117663  0.140983  -0.006097  ... -0.267632  0.466206   \n",
       "GSM6745600 -0.235702  -1.159822 -0.267197  -1.511998  ...  0.007233  1.309818   \n",
       "GSM6745601 -0.235702   1.760144  1.722682   0.688935  ...  0.969260  2.153430   \n",
       "GSM6745602 -0.235702   1.121401  0.498141   0.225580  ...  0.282098  2.715838   \n",
       "GSM6745603 -0.235702   0.391410 -0.165152   0.804773  ...  0.831828  1.028614   \n",
       "GSM6745604 -0.235702   0.756406  0.804276  -0.121935  ...  1.381557  0.185003   \n",
       "GSM6745605 -0.235702   0.026414 -0.012084   0.225580  ...  2.206152  0.466206   \n",
       "GSM6745606 -0.235702  -1.159822 -0.828445  -1.048644  ... -0.817361 -0.096201   \n",
       "GSM6745607 -0.235702   0.208912 -0.012084   0.457258  ... -0.130199  0.185003   \n",
       "GSM6745609 -0.235702  -0.886075 -1.185602   0.109742  ... -1.367091 -0.658609   \n",
       "GSM6745610 -0.235702   0.117663 -0.267197   0.109742  ... -0.542496 -0.658609   \n",
       "GSM6745611 -0.235702   0.117663 -0.216174  -0.816967  ... -0.405064 -0.658609   \n",
       "GSM6745612 -0.235702  -0.156084  1.161434  -0.932805  ... -0.542496  0.185003   \n",
       "GSM6745613 -0.235702   2.033890  0.804276   2.078998  ... -0.130199 -0.939813   \n",
       "GSM6745614 -0.235702   0.026414 -0.522310  -0.585290  ... -0.267632 -0.658609   \n",
       "GSM6745615  4.242641   2.672633  3.508470   2.194836  ...  1.518990 -0.096201   \n",
       "GSM6745616  4.242641   0.391410  0.345073   0.225580  ... -0.267632 -0.096201   \n",
       "GSM6745617 -0.235702   0.117663  0.396096  -0.932805  ...  0.007233 -1.502221   \n",
       "GSM6745618 -0.235702   0.208912  0.855299  -0.932805  ...  0.556963  1.309818   \n",
       "GSM6745619 -0.235702  -1.889813 -1.032535  -1.859514  ... -1.229658 -1.221017   \n",
       "GSM6745620 -0.235702  -0.338582 -0.114129   0.341419  ... -0.954794 -0.658609   \n",
       "GSM6745621 -0.235702  -0.521079 -0.522310  -0.701128  ... -0.954794  0.185003   \n",
       "GSM6745622 -0.235702   0.391410  0.447118   0.457258  ...  0.144666 -0.377405   \n",
       "GSM6745623 -0.235702  -0.703577  1.161434   0.225580  ... -0.542496 -0.377405   \n",
       "GSM6745624 -0.235702  -0.612328 -0.216174  -0.237774  ...  0.282098 -0.096201   \n",
       "GSM6745625 -0.235702  -0.977324 -0.267197  -0.469451  ... -0.542496 -0.658609   \n",
       "GSM6745626 -0.235702   0.482659  1.212456   1.731482  ...  1.381557  0.466206   \n",
       "GSM6745627 -0.235702  -1.798564 -2.359120  -1.975353  ... -1.916820 -1.502221   \n",
       "GSM6745628 -0.235702  -0.886075 -0.114129   0.109742  ...  1.656422 -0.377405   \n",
       "GSM6745629 -0.235702   0.391410  0.345073  -0.353613  ... -0.817361 -0.658609   \n",
       "GSM6745630 -0.235702  -0.794826 -1.593783  -0.932805  ... -1.229658 -0.939813   \n",
       "GSM6745631 -0.235702   0.665157 -0.165152  -0.006097  ... -0.542496 -0.658609   \n",
       "GSM6745632 -0.235702   2.216388 -1.134580   1.615643  ...  0.282098  0.185003   \n",
       "GSM6745633 -0.235702  -0.064835  0.294051   1.615643  ...  2.618449  2.434634   \n",
       "GSM6745634 -0.235702  -0.338582 -0.471287  -1.048644  ... -0.817361 -1.502221   \n",
       "GSM6745635 -0.235702  -0.886075 -0.879467  -0.237774  ... -0.405064 -0.096201   \n",
       "GSM6745636 -0.235702  -0.338582 -0.369242   1.152289  ...  0.831828  0.466206   \n",
       "GSM6745637 -0.235702  -0.703577 -0.981512   0.341419  ... -0.267632  0.747410   \n",
       "\n",
       "GeneID          4540      4541      4556      4519      4576      4571  \\\n",
       "GSM6745599 -0.318275 -0.133587  0.136065 -0.604440 -0.270274 -0.375282   \n",
       "GSM6745600 -0.981598 -0.819478 -0.337051 -1.184536 -1.084285 -0.450273   \n",
       "GSM6745601  1.484606  1.848533  1.741280  0.865277  1.357749  1.874436   \n",
       "GSM6745602  0.638045  1.031012  0.913327  0.141225  0.041475  0.237141   \n",
       "GSM6745603  0.434101  0.974346  1.293509  0.187012  0.110753  0.624593   \n",
       "GSM6745604  1.132972  1.712771  2.053874  0.594337  1.115277  1.137029   \n",
       "GSM6745605 -0.349242  0.224115  0.786600 -0.666955 -0.339551  0.862063   \n",
       "GSM6745606 -0.984529 -0.787604 -0.413087 -1.281728 -0.997688 -0.900216   \n",
       "GSM6745607  0.108853  0.397654  0.474005 -0.268787 -0.443467 -0.125313   \n",
       "GSM6745609 -0.198254 -0.324244 -0.945342 -0.949860 -1.188201 -0.950210   \n",
       "GSM6745610  0.579042  0.754766  0.575387 -0.268543 -0.339551 -0.400279   \n",
       "GSM6745611 -0.861943 -0.637086 -0.522918 -0.602974 -0.374190 -0.862721   \n",
       "GSM6745612 -0.479158 -0.164281  0.017786 -0.264147 -0.304912 -0.637749   \n",
       "GSM6745613 -0.401282 -0.694342 -1.156555 -0.068665 -0.149038  0.124655   \n",
       "GSM6745614 -0.542559 -0.396846 -0.092044 -0.319092 -0.478106 -0.362783   \n",
       "GSM6745615  2.198502  1.854436  1.741280  2.608864  1.929288  1.624467   \n",
       "GSM6745616  0.107754 -0.046818 -0.235669  0.433776  0.180030  0.249640   \n",
       "GSM6745617 -0.400183 -0.797048 -0.928445  0.090553  0.439821 -0.162809   \n",
       "GSM6745618  0.920782  1.065838  1.031606  1.194703  0.872806  0.537104   \n",
       "GSM6745619 -1.899621 -1.863662 -1.587428 -1.646929 -1.569228 -1.712614   \n",
       "GSM6745620 -0.717918 -0.702015 -0.506021 -0.409690 -0.582023 -0.850222   \n",
       "GSM6745621 -0.593133 -0.405701 -0.320154 -0.369275 -0.495426 -0.662746   \n",
       "GSM6745622 -0.089227 -0.238064 -0.320154  0.131579  0.024156 -0.450273   \n",
       "GSM6745623  0.290442  0.295538  0.203653  0.197757 -0.114399 -0.300291   \n",
       "GSM6745624  0.569148  0.543450  0.693666  0.771382  0.803528  0.462113   \n",
       "GSM6745625 -0.102054 -0.185531 -0.269463  0.247574  0.110753  0.162151   \n",
       "GSM6745626  1.629181  1.468400  1.563861  1.587498  1.808053  1.699458   \n",
       "GSM6745627 -2.095136 -1.896718 -1.511392 -1.923485 -1.655824 -1.825100   \n",
       "GSM6745628 -0.251393 -1.024301 -1.384664  0.168819  0.526418  0.762076   \n",
       "GSM6745629 -0.627032 -0.993607 -1.190349 -0.302731 -0.616661 -0.862721   \n",
       "GSM6745630 -1.498146 -1.513043 -1.317076 -1.254255 -1.240159 -1.350160   \n",
       "GSM6745631  0.576111  0.642025  0.685218  0.578953  0.509099  0.262138   \n",
       "GSM6745632  1.386757  1.317292  1.369546  0.938049  1.201874  1.474486   \n",
       "GSM6745633  2.424618  1.288369 -0.218772  3.018998  3.245562  2.811819   \n",
       "GSM6745634 -0.806605 -1.298776 -1.604325 -0.500777 -0.720578 -0.537762   \n",
       "GSM6745635 -0.742472 -0.869061 -0.801718 -0.835331 -0.893771 -1.000204   \n",
       "GSM6745636  0.280914  0.466125  0.490902  0.140248 -0.374190 -0.150310   \n",
       "GSM6745637  0.177934 -0.092858 -0.108941 -0.174404 -0.045122  0.024668   \n",
       "\n",
       "GeneID      Edited (1) or Unedited (0)  editing mechanism  \n",
       "GSM6745599                           1                BE4  \n",
       "GSM6745600                           1                BE4  \n",
       "GSM6745601                           1                BE4  \n",
       "GSM6745602                           1               ABE8  \n",
       "GSM6745603                           1               ABE8  \n",
       "GSM6745604                           1               ABE8  \n",
       "GSM6745605                           1               Cas9  \n",
       "GSM6745606                           1               Cas9  \n",
       "GSM6745607                           1               Cas9  \n",
       "GSM6745609                           1          Utelectro  \n",
       "GSM6745610                           1          Utelectro  \n",
       "GSM6745611                           1                BE4  \n",
       "GSM6745612                           1                BE4  \n",
       "GSM6745613                           1                BE4  \n",
       "GSM6745614                           1               ABE8  \n",
       "GSM6745615                           1               ABE8  \n",
       "GSM6745616                           1               ABE8  \n",
       "GSM6745617                           1               Cas9  \n",
       "GSM6745618                           1               Cas9  \n",
       "GSM6745619                           1               Cas9  \n",
       "GSM6745620                           1          Utelectro  \n",
       "GSM6745621                           1          Utelectro  \n",
       "GSM6745622                           1          Utelectro  \n",
       "GSM6745623                           1              dCas9  \n",
       "GSM6745624                           1              dCas9  \n",
       "GSM6745625                           1              dCas9  \n",
       "GSM6745626                           1           BE4alone  \n",
       "GSM6745627                           1           BE4alone  \n",
       "GSM6745628                           1           BE4alone  \n",
       "GSM6745629                           1          ABE8alone  \n",
       "GSM6745630                           1          ABE8alone  \n",
       "GSM6745631                           1          ABE8alone  \n",
       "GSM6745632                           0                 UT  \n",
       "GSM6745633                           0                 UT  \n",
       "GSM6745634                           0                 UT  \n",
       "GSM6745635                           0                 UT  \n",
       "GSM6745636                           0                 UT  \n",
       "GSM6745637                           0                 UT  \n",
       "\n",
       "[38 rows x 39378 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.decomposition import PCA\n",
    "import numpy as np\n",
    "\n",
    "df = pd.read_csv('../../data/GSE218462_raw_counts_GRCh38.p13_NCBI.tsv', sep='\\t')\n",
    "df = df.T\n",
    "df.columns = df.iloc[0]\n",
    "df = df[1:]\n",
    "\n",
    "scaler = StandardScaler()\n",
    "scaled_data = scaler.fit_transform(df)\n",
    "scaled_df = pd.DataFrame(scaled_data)\n",
    "\n",
    "scaled_df.columns = df.columns\n",
    "scaled_df.index = df.index\n",
    "unedited = ['GSM6745632', 'GSM6745633', 'GSM6745634', 'GSM6745635', 'GSM6745636', 'GSM6745637']\n",
    "scaled_df['Edited (1) or Unedited (0)'] = scaled_df.index.map(lambda gene: 0 if gene in unedited else 1)\n",
    "mechanisms = {\n",
    "    \"BE4\": [\"GSM6745599\", \"GSM6745600\", \"GSM6745601\", \"GSM6745611\", \"GSM6745612\", \"GSM6745613\"],\n",
    "    \"ABE8\": [\"GSM6745602\", \"GSM6745603\", \"GSM6745604\", \"GSM6745614\", \"GSM6745615\", \"GSM6745616\"],\n",
    "    \"Cas9\": [\"GSM6745605\", \"GSM6745606\", \"GSM6745607\", \"GSM6745617\", \"GSM6745618\", \"GSM6745619\"],\n",
    "    \"Utelectro\": [\"GSM6745608\", \"GSM6745609\", \"GSM6745610\", \"GSM6745620\", \"GSM6745621\", \"GSM6745622\"],\n",
    "    \"dCas9\": [\"GSM6745623\", \"GSM6745624\", \"GSM6745625\"],\n",
    "    \"BE4alone\": [\"GSM6745626\", \"GSM6745627\", \"GSM6745628\"],\n",
    "    \"ABE8alone\": [\"GSM6745629\", \"GSM6745630\", \"GSM6745631\"],\n",
    "    \"UT\": [\"GSM6745632\", \"GSM6745633\", \"GSM6745634\", \"GSM6745635\", \"GSM6745636\", \"GSM6745637\"]\n",
    "}\n",
    "\n",
    "# Inverting the dictionary to map gene code to its corresponding key\n",
    "mechanism_map = {gene: mechanism for mechanism, genes in mechanisms.items() for gene in genes}\n",
    "\n",
    "# Adding a new column \"editing mechanism\" to categorize the gene codes in the index\n",
    "scaled_df['editing mechanism'] = scaled_df.index.map(mechanism_map)\n",
    "\n",
    "scaled_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/n1/ddw6p35j0310zh6cql00y29r0000gn/T/ipykernel_5877/3075730881.py:2: DtypeWarning: Columns (8,9) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  tsv_df = pd.read_csv(tsv_file_path, sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>Description</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>GeneType</th>\n",
       "      <th>EnsemblGeneID</th>\n",
       "      <th>Status</th>\n",
       "      <th>ChrAcc</th>\n",
       "      <th>ChrStart</th>\n",
       "      <th>ChrStop</th>\n",
       "      <th>Orientation</th>\n",
       "      <th>Length</th>\n",
       "      <th>GOFunctionID</th>\n",
       "      <th>GOProcessID</th>\n",
       "      <th>GOComponentID</th>\n",
       "      <th>GOFunction</th>\n",
       "      <th>GOProcess</th>\n",
       "      <th>GOComponent</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>100287102</td>\n",
       "      <td>DDX11L1</td>\n",
       "      <td>DEAD/H-box helicase 11 like 1 (pseudogene)</td>\n",
       "      <td>NaN</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>ENSG00000290825</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>11874</td>\n",
       "      <td>14409</td>\n",
       "      <td>positive</td>\n",
       "      <td>1652</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>653635</td>\n",
       "      <td>WASH7P</td>\n",
       "      <td>WASP family homolog 7, pseudogene</td>\n",
       "      <td>FAM39F|WASH5P</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>14362</td>\n",
       "      <td>29370</td>\n",
       "      <td>negative</td>\n",
       "      <td>1769</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>102466751</td>\n",
       "      <td>MIR6859-1</td>\n",
       "      <td>microRNA 6859-1</td>\n",
       "      <td>hsa-mir-6859-1</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>ENSG00000278267</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>17369</td>\n",
       "      <td>17436</td>\n",
       "      <td>negative</td>\n",
       "      <td>68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>107985730</td>\n",
       "      <td>MIR1302-2HG</td>\n",
       "      <td>MIR1302-2 host gene</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>29926</td>\n",
       "      <td>31295</td>\n",
       "      <td>positive</td>\n",
       "      <td>538</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>100302278</td>\n",
       "      <td>MIR1302-2</td>\n",
       "      <td>microRNA 1302-2</td>\n",
       "      <td>MIRN1302-2|hsa-mir-1302-2</td>\n",
       "      <td>ncRNA</td>\n",
       "      <td>ENSG00000284332</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>30366</td>\n",
       "      <td>30503</td>\n",
       "      <td>positive</td>\n",
       "      <td>138</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GO:0035195</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>miRNA-mediated gene silencing</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39371</th>\n",
       "      <td>4541</td>\n",
       "      <td>ND6</td>\n",
       "      <td>NADH dehydrogenase subunit 6</td>\n",
       "      <td>MTND6</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>14149</td>\n",
       "      <td>14673</td>\n",
       "      <td>negative</td>\n",
       "      <td>525</td>\n",
       "      <td>GO:0008137</td>\n",
       "      <td>GO:0006120///GO:0009060///GO:0032981///GO:0035...</td>\n",
       "      <td>GO:0005739///GO:0005743///GO:0005747</td>\n",
       "      <td>NADH dehydrogenase (ubiquinone) activity</td>\n",
       "      <td>mitochondrial electron transport, NADH to ubiq...</td>\n",
       "      <td>mitochondrion///mitochondrial inner membrane//...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39372</th>\n",
       "      <td>4556</td>\n",
       "      <td>TRNE</td>\n",
       "      <td>tRNA-Glu</td>\n",
       "      <td>MTTE</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>14674</td>\n",
       "      <td>14742</td>\n",
       "      <td>negative</td>\n",
       "      <td>69</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39373</th>\n",
       "      <td>4519</td>\n",
       "      <td>CYTB</td>\n",
       "      <td>cytochrome b</td>\n",
       "      <td>MTCYB</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>14747</td>\n",
       "      <td>15887</td>\n",
       "      <td>positive</td>\n",
       "      <td>1141</td>\n",
       "      <td>GO:0008121///GO:0046872</td>\n",
       "      <td>GO:0006122///GO:0045333///GO:1902600</td>\n",
       "      <td>GO:0005739///GO:0005743///GO:0005750///GO:0016020</td>\n",
       "      <td>ubiquinol-cytochrome-c reductase activity///me...</td>\n",
       "      <td>mitochondrial electron transport, ubiquinol to...</td>\n",
       "      <td>mitochondrion///mitochondrial inner membrane//...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39374</th>\n",
       "      <td>4576</td>\n",
       "      <td>TRNT</td>\n",
       "      <td>tRNA-Thr</td>\n",
       "      <td>MTTT</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>positive</td>\n",
       "      <td>66</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39375</th>\n",
       "      <td>4571</td>\n",
       "      <td>TRNP</td>\n",
       "      <td>tRNA-Pro</td>\n",
       "      <td>MTTP</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>NaN</td>\n",
       "      <td>active</td>\n",
       "      <td>NC_012920.1</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>negative</td>\n",
       "      <td>68</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>39376 rows Ã— 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          GeneID       Symbol                                 Description  \\\n",
       "0      100287102      DDX11L1  DEAD/H-box helicase 11 like 1 (pseudogene)   \n",
       "1         653635       WASH7P           WASP family homolog 7, pseudogene   \n",
       "2      102466751    MIR6859-1                             microRNA 6859-1   \n",
       "3      107985730  MIR1302-2HG                         MIR1302-2 host gene   \n",
       "4      100302278    MIR1302-2                             microRNA 1302-2   \n",
       "...          ...          ...                                         ...   \n",
       "39371       4541          ND6                NADH dehydrogenase subunit 6   \n",
       "39372       4556         TRNE                                    tRNA-Glu   \n",
       "39373       4519         CYTB                                cytochrome b   \n",
       "39374       4576         TRNT                                    tRNA-Thr   \n",
       "39375       4571         TRNP                                    tRNA-Pro   \n",
       "\n",
       "                        Synonyms        GeneType    EnsemblGeneID  Status  \\\n",
       "0                            NaN          pseudo  ENSG00000290825  active   \n",
       "1                  FAM39F|WASH5P          pseudo              NaN  active   \n",
       "2                 hsa-mir-6859-1           ncRNA  ENSG00000278267  active   \n",
       "3                            NaN           ncRNA              NaN  active   \n",
       "4      MIRN1302-2|hsa-mir-1302-2           ncRNA  ENSG00000284332  active   \n",
       "...                          ...             ...              ...     ...   \n",
       "39371                      MTND6  protein-coding              NaN  active   \n",
       "39372                       MTTE            tRNA              NaN  active   \n",
       "39373                      MTCYB  protein-coding              NaN  active   \n",
       "39374                       MTTT            tRNA              NaN  active   \n",
       "39375                       MTTP            tRNA              NaN  active   \n",
       "\n",
       "             ChrAcc ChrStart ChrStop Orientation  Length  \\\n",
       "0      NC_000001.11    11874   14409    positive    1652   \n",
       "1      NC_000001.11    14362   29370    negative    1769   \n",
       "2      NC_000001.11    17369   17436    negative      68   \n",
       "3      NC_000001.11    29926   31295    positive     538   \n",
       "4      NC_000001.11    30366   30503    positive     138   \n",
       "...             ...      ...     ...         ...     ...   \n",
       "39371   NC_012920.1    14149   14673    negative     525   \n",
       "39372   NC_012920.1    14674   14742    negative      69   \n",
       "39373   NC_012920.1    14747   15887    positive    1141   \n",
       "39374   NC_012920.1    15888   15953    positive      66   \n",
       "39375   NC_012920.1    15956   16023    negative      68   \n",
       "\n",
       "                  GOFunctionID  \\\n",
       "0                          NaN   \n",
       "1                          NaN   \n",
       "2                          NaN   \n",
       "3                          NaN   \n",
       "4                          NaN   \n",
       "...                        ...   \n",
       "39371               GO:0008137   \n",
       "39372                      NaN   \n",
       "39373  GO:0008121///GO:0046872   \n",
       "39374                      NaN   \n",
       "39375                      NaN   \n",
       "\n",
       "                                             GOProcessID  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                                             GO:0035195   \n",
       "...                                                  ...   \n",
       "39371  GO:0006120///GO:0009060///GO:0032981///GO:0035...   \n",
       "39372                                                NaN   \n",
       "39373               GO:0006122///GO:0045333///GO:1902600   \n",
       "39374                                                NaN   \n",
       "39375                                                NaN   \n",
       "\n",
       "                                           GOComponentID  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                                                    NaN   \n",
       "...                                                  ...   \n",
       "39371               GO:0005739///GO:0005743///GO:0005747   \n",
       "39372                                                NaN   \n",
       "39373  GO:0005739///GO:0005743///GO:0005750///GO:0016020   \n",
       "39374                                                NaN   \n",
       "39375                                                NaN   \n",
       "\n",
       "                                              GOFunction  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                                                    NaN   \n",
       "...                                                  ...   \n",
       "39371           NADH dehydrogenase (ubiquinone) activity   \n",
       "39372                                                NaN   \n",
       "39373  ubiquinol-cytochrome-c reductase activity///me...   \n",
       "39374                                                NaN   \n",
       "39375                                                NaN   \n",
       "\n",
       "                                               GOProcess  \\\n",
       "0                                                    NaN   \n",
       "1                                                    NaN   \n",
       "2                                                    NaN   \n",
       "3                                                    NaN   \n",
       "4                          miRNA-mediated gene silencing   \n",
       "...                                                  ...   \n",
       "39371  mitochondrial electron transport, NADH to ubiq...   \n",
       "39372                                                NaN   \n",
       "39373  mitochondrial electron transport, ubiquinol to...   \n",
       "39374                                                NaN   \n",
       "39375                                                NaN   \n",
       "\n",
       "                                             GOComponent  \n",
       "0                                                    NaN  \n",
       "1                                                    NaN  \n",
       "2                                                    NaN  \n",
       "3                                                    NaN  \n",
       "4                                                    NaN  \n",
       "...                                                  ...  \n",
       "39371  mitochondrion///mitochondrial inner membrane//...  \n",
       "39372                                                NaN  \n",
       "39373  mitochondrion///mitochondrial inner membrane//...  \n",
       "39374                                                NaN  \n",
       "39375                                                NaN  \n",
       "\n",
       "[39376 rows x 18 columns]"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tsv_file_path = '../../data/Human.GRCh38.p13.annot.tsv'  # Replace with the actual path to your TSV file\n",
    "tsv_df = pd.read_csv(tsv_file_path, sep='\\t')\n",
    "\n",
    "tsv_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/n1/ddw6p35j0310zh6cql00y29r0000gn/T/ipykernel_5877/3509068322.py:2: DtypeWarning: Columns (8,9) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  metadata = pd.read_csv(metadata_file_path, sep='\\t')\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merged Data (first few rows):\n",
      "Index(['GSM6745599', 'GSM6745600', 'GSM6745601', 'GSM6745602', 'GSM6745603',\n",
      "       'GSM6745604', 'GSM6745605', 'GSM6745606', 'GSM6745607', 'GSM6745609',\n",
      "       'GSM6745610', 'GSM6745611', 'GSM6745612', 'GSM6745613', 'GSM6745614',\n",
      "       'GSM6745615', 'GSM6745616', 'GSM6745617', 'GSM6745618', 'GSM6745619',\n",
      "       'GSM6745620', 'GSM6745621', 'GSM6745622', 'GSM6745623', 'GSM6745624',\n",
      "       'GSM6745625', 'GSM6745626', 'GSM6745627', 'GSM6745628', 'GSM6745629',\n",
      "       'GSM6745630', 'GSM6745631', 'GSM6745632', 'GSM6745633', 'GSM6745634',\n",
      "       'GSM6745635', 'GSM6745636', 'GSM6745637', 'Symbol', 'Description',\n",
      "       'Synonyms', 'GeneType', 'EnsemblGeneID', 'Status', 'ChrAcc', 'ChrStart',\n",
      "       'ChrStop', 'Orientation', 'Length', 'GOFunctionID', 'GOProcessID',\n",
      "       'GOComponentID', 'GOFunction', 'GOProcess', 'GOComponent'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "metadata_file_path = \"../../data/Human.GRCh38.p13.annot.tsv\"\n",
    "metadata = pd.read_csv(metadata_file_path, sep='\\t')\n",
    "\n",
    "transposed_data = scaled_df.T\n",
    "transposed_data = transposed_data.reset_index()\n",
    "transposed_data.columns.values[0] = 'GeneID'\n",
    "\n",
    "\n",
    "merged_data = transposed_data.merge(metadata, on='GeneID', how='left').set_index('GeneID')\n",
    "merged_data = merged_data.iloc[:39378]\n",
    "\n",
    "print(\"Merged Data (first few rows):\")\n",
    "\n",
    "# Optionally, save to a new file\n",
    "output_file_path = \"merged_gene_expression_with_metadata.csv\"\n",
    "merged_data.to_csv(output_file_path)\n",
    "print(merged_data.columns)\n",
    "# print(f\"Merged data saved to {output_file_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "merged_data = merged_data.iloc[:, :-6]\n",
    "merged_data = merged_data.drop(columns=['EnsemblGeneID'])\n",
    "merged_data\n",
    "output_file_path = \"testing.csv\"\n",
    "merged_data.to_csv(output_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1319"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merged_data_nulls = merged_data['Description'].isnull().sum()\n",
    "merged_data_nulls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of rows: 38059, Number of columns: 48\n",
      "          GeneID GSM6745599 GSM6745600 GSM6745601 GSM6745602 GSM6745603  \\\n",
      "0      100287102   0.212625  -1.311857   1.127315   0.365073    0.66997   \n",
      "1         653635  -0.318248  -1.500716    1.33452  -0.318248    0.11174   \n",
      "2      102466751  -0.042237  -1.647234   0.492762  -0.577236  -0.042237   \n",
      "3      107985730   0.184916  -0.818915  -0.818915   0.184916   1.188747   \n",
      "4      100302278    1.06341    1.06341  -0.773389  -0.773389  -0.773389   \n",
      "...          ...        ...        ...        ...        ...        ...   \n",
      "38054       4541  -0.133587  -0.819478   1.848533   1.031012   0.974346   \n",
      "38055       4556   0.136065  -0.337051    1.74128   0.913327   1.293509   \n",
      "38056       4519   -0.60444  -1.184536   0.865277   0.141225   0.187012   \n",
      "38057       4576  -0.270274  -1.084285   1.357749   0.041475   0.110753   \n",
      "38058       4571  -0.375282  -0.450273   1.874436   0.237141   0.624593   \n",
      "\n",
      "      GSM6745604 GSM6745605 GSM6745606 GSM6745607  ...       Symbol  \\\n",
      "0      -0.092271    0.66997  -0.702064   0.517522  ...      DDX11L1   \n",
      "1       0.622352   0.622352  -1.231973   0.138615  ...       WASH7P   \n",
      "2        1.56276  -1.112235  -0.577236    1.56276  ...    MIR6859-1   \n",
      "3       0.184916  -0.818915  -1.822745   1.188747  ...  MIR1302-2HG   \n",
      "4      -0.773389  -0.773389    1.06341  -0.773389  ...    MIR1302-2   \n",
      "...          ...        ...        ...        ...  ...          ...   \n",
      "38054   1.712771   0.224115  -0.787604   0.397654  ...          ND6   \n",
      "38055   2.053874     0.7866  -0.413087   0.474005  ...         TRNE   \n",
      "38056   0.594337  -0.666955  -1.281728  -0.268787  ...         CYTB   \n",
      "38057   1.115277  -0.339551  -0.997688  -0.443467  ...         TRNT   \n",
      "38058   1.137029   0.862063  -0.900216  -0.125313  ...         TRNP   \n",
      "\n",
      "                                      Description                   Synonyms  \\\n",
      "0      DEAD/H-box helicase 11 like 1 (pseudogene)                        NaN   \n",
      "1               WASP family homolog 7, pseudogene              FAM39F|WASH5P   \n",
      "2                                 microRNA 6859-1             hsa-mir-6859-1   \n",
      "3                             MIR1302-2 host gene                        NaN   \n",
      "4                                 microRNA 1302-2  MIRN1302-2|hsa-mir-1302-2   \n",
      "...                                           ...                        ...   \n",
      "38054                NADH dehydrogenase subunit 6                      MTND6   \n",
      "38055                                    tRNA-Glu                       MTTE   \n",
      "38056                                cytochrome b                      MTCYB   \n",
      "38057                                    tRNA-Thr                       MTTT   \n",
      "38058                                    tRNA-Pro                       MTTP   \n",
      "\n",
      "             GeneType  Status        ChrAcc ChrStart ChrStop Orientation  \\\n",
      "0              pseudo  active  NC_000001.11    11874   14409    positive   \n",
      "1              pseudo  active  NC_000001.11    14362   29370    negative   \n",
      "2               ncRNA  active  NC_000001.11    17369   17436    negative   \n",
      "3               ncRNA  active  NC_000001.11    29926   31295    positive   \n",
      "4               ncRNA  active  NC_000001.11    30366   30503    positive   \n",
      "...               ...     ...           ...      ...     ...         ...   \n",
      "38054  protein-coding  active   NC_012920.1    14149   14673    negative   \n",
      "38055            tRNA  active   NC_012920.1    14674   14742    negative   \n",
      "38056  protein-coding  active   NC_012920.1    14747   15887    positive   \n",
      "38057            tRNA  active   NC_012920.1    15888   15953    positive   \n",
      "38058            tRNA  active   NC_012920.1    15956   16023    negative   \n",
      "\n",
      "       Length  \n",
      "0      1652.0  \n",
      "1      1769.0  \n",
      "2        68.0  \n",
      "3       538.0  \n",
      "4       138.0  \n",
      "...       ...  \n",
      "38054   525.0  \n",
      "38055    69.0  \n",
      "38056  1141.0  \n",
      "38057    66.0  \n",
      "38058    68.0  \n",
      "\n",
      "[38059 rows x 49 columns]\n"
     ]
    }
   ],
   "source": [
    "# Delete any genes with missing descriptions\n",
    "merged_data = merged_data.dropna(subset=['Description'])\n",
    "print(f\"Number of rows: {merged_data.shape[0]}, Number of columns: {merged_data.shape[1]}\")\n",
    "merged_data.reset_index(inplace=True) # Apparently, I had GeneID as the index, so I reset it to a column cuz it's easier to work with\n",
    "print(merged_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Relevant genes with contributions for PC1:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "105378947  0.508363         11.882389\n",
      "729759     0.375923          8.786761\n",
      "643837     0.374651          8.757042\n",
      "100133331  0.287999          6.731646\n",
      "102466751  0.287508          6.720178\n",
      "79854      0.257980          6.029991\n",
      "284600     0.174607          4.081236\n",
      "653635     0.165884          3.877335\n",
      "107984847  0.152373          3.561533\n",
      "102465909  0.134157          3.135772\n",
      "100996442  0.124959          2.920774\n",
      "26155      0.123478          2.886151\n",
      "81399      0.112900          2.638916\n",
      "79501      0.110382          2.580052\n",
      "284593     0.104132          2.433966\n",
      "107984841  0.097564          2.280439\n",
      "100132287  0.089612          2.094587\n",
      "645520     0.087955          2.055835\n",
      "112268260  0.078926          1.844806\n",
      "102723897  0.077772          1.817834\n",
      "729737     0.071377          1.668359\n",
      "\n",
      "Relevant genes with contributions for PC2:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "102466751  0.346415          6.111373\n",
      "105378947  0.247890          4.373217\n",
      "79501      0.244025          4.305036\n",
      "645520     0.243838          4.301731\n",
      "107985730  0.222422          3.923922\n",
      "100302278  0.221497          3.907591\n",
      "102723897  0.221339          3.904803\n",
      "729737     0.208153          3.672176\n",
      "100287102  0.203459          3.589378\n",
      "729759     0.194288          3.427580\n",
      "84808      0.182840          3.225616\n",
      "79854      0.180821          3.189997\n",
      "102725121  0.176105          3.106798\n",
      "84069      0.173792          3.066004\n",
      "107985728  0.153304          2.704559\n",
      "81399      0.153024          2.699605\n",
      "148398     0.145904          2.574010\n",
      "339451     0.140762          2.483286\n",
      "107984849  0.139017          2.452510\n",
      "653635     0.130292          2.298578\n",
      "100288069  0.120894          2.132781\n",
      "100996442  0.119246          2.103708\n",
      "100130417  0.116094          2.048107\n",
      "113219467  0.115856          2.043900\n",
      "26155      0.114792          2.025132\n",
      "107984850  0.113915          2.009661\n",
      "101928626  0.112847          1.990825\n",
      "105378580  0.105276          1.857258\n",
      "400728     0.104125          1.836944\n",
      "107984841  0.098384          1.735670\n",
      "\n",
      "Relevant genes with contributions for PC3:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "107985728  0.514852         11.412812\n",
      "105378947  0.335987          7.447883\n",
      "26155      0.318922          7.069606\n",
      "148398     0.298578          6.618632\n",
      "729759     0.284859          6.314518\n",
      "84069      0.255956          5.673823\n",
      "100288069  0.205312          4.551194\n",
      "84808      0.204061          4.523468\n",
      "107985730  0.166808          3.697670\n",
      "339451     0.149365          3.311000\n",
      "101928626  0.146040          3.237306\n",
      "112268260  0.132412          2.935211\n",
      "79501      0.123465          2.736882\n",
      "284593     0.109082          2.418043\n",
      "102466751  0.102962          2.282370\n",
      "105378580  0.097188          2.154394\n",
      "113219467  0.093237          2.066792\n",
      "81399      0.087273          1.934587\n",
      "100132287  0.086024          1.906907\n",
      "100130417  0.076996          1.706784\n",
      "653635     0.064629          1.432638\n",
      "107984841  0.063807          1.414420\n",
      "79854      0.062436          1.384039\n",
      "729737     0.058791          1.303230\n",
      "\n",
      "Relevant genes with contributions for PC4:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "729759     0.600189         14.644901\n",
      "107985728  0.377479          9.210670\n",
      "105378947  0.370559          9.041835\n",
      "84808      0.266800          6.510060\n",
      "102725121  0.233192          5.689998\n",
      "148398     0.178477          4.354930\n",
      "79501      0.171874          4.193820\n",
      "79854      0.155159          3.785966\n",
      "107985730  0.152102          3.711375\n",
      "107984850  0.142694          3.481815\n",
      "102466751  0.124771          3.044480\n",
      "400728     0.108440          2.645996\n",
      "26155      0.106406          2.596352\n",
      "284600     0.093330          2.277308\n",
      "100287102  0.090143          2.199533\n",
      "84069      0.086447          2.109351\n",
      "100288069  0.083837          2.045659\n",
      "102465909  0.074250          1.811741\n",
      "100133331  0.063343          1.545605\n",
      "101928626  0.063156          1.541044\n",
      "100996442  0.059592          1.454071\n",
      "339451     0.056371          1.375487\n",
      "\n",
      "Relevant genes with contributions for PC5:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "84808      0.585445         19.368596\n",
      "107985728  0.531357         17.579171\n",
      "102466751  0.480133         15.884499\n",
      "84069      0.222784          7.370484\n",
      "339451     0.162616          5.379922\n",
      "105378947  0.160878          5.322417\n",
      "102723897  0.092404          3.057049\n",
      "107985730  0.071219          2.356174\n",
      "26155      0.068839          2.277447\n",
      "148398     0.066158          2.188746\n",
      "107984849  0.061783          2.044014\n",
      "79501      0.045305          1.498834\n",
      "653635     0.042878          1.418554\n",
      "729759     0.039180          1.296201\n",
      "102725121  0.038192          1.263535\n",
      "112268260  0.037935          1.255014\n",
      "\n",
      "Relevant genes with contributions for PC6:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "102725121  0.767866         23.977469\n",
      "102466751  0.331770         10.359892\n",
      "107985728  0.312459          9.756879\n",
      "105378947  0.262474          8.196034\n",
      "102723897  0.210232          6.564728\n",
      "148398     0.163527          5.106327\n",
      "26155      0.101092          3.156707\n",
      "107984849  0.085434          2.667761\n",
      "81399      0.084832          2.648982\n",
      "645520     0.078041          2.436913\n",
      "100302278  0.067762          2.115940\n",
      "113219467  0.062604          1.954893\n",
      "729759     0.056995          1.779736\n",
      "107984847  0.052597          1.642407\n",
      "107985730  0.051339          1.603104\n",
      "79501      0.049702          1.552009\n",
      "79854      0.045240          1.412668\n",
      "729737     0.037246          1.163041\n",
      "100132287  0.034705          1.083688\n",
      "\n",
      "Relevant genes with contributions for PC7:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "26155      0.787967         24.810815\n",
      "107985728  0.325891         10.261380\n",
      "102725121  0.273257          8.604090\n",
      "102466751  0.237898          7.490734\n",
      "84808      0.204925          6.452488\n",
      "105378947  0.163370          5.144050\n",
      "79854      0.147025          4.629410\n",
      "148398     0.110155          3.468479\n",
      "100130417  0.075891          2.389598\n",
      "645520     0.059521          1.874145\n",
      "81399      0.057647          1.815137\n",
      "107984850  0.057028          1.795648\n",
      "100288069  0.054297          1.709650\n",
      "102723897  0.050138          1.578716\n",
      "729737     0.048651          1.531880\n",
      "100302278  0.044770          1.409686\n",
      "84069      0.044161          1.390506\n",
      "339451     0.040164          1.264639\n",
      "400728     0.040132          1.263631\n",
      "\n",
      "Relevant genes with contributions for PC8:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "79854      0.561874         13.561620\n",
      "729759     0.475868         11.485749\n",
      "107985730  0.301702          7.282008\n",
      "102725121  0.238134          5.747694\n",
      "105378947  0.214393          5.174684\n",
      "102466751  0.212976          5.140470\n",
      "81399      0.201051          4.852638\n",
      "107984850  0.199588          4.817348\n",
      "148398     0.161223          3.891339\n",
      "107985728  0.138967          3.354157\n",
      "100130417  0.135480          3.269989\n",
      "400728     0.100798          2.432908\n",
      "79501      0.080842          1.951239\n",
      "645520     0.077865          1.879392\n",
      "100132287  0.073429          1.772316\n",
      "653635     0.066768          1.611544\n",
      "100302278  0.062312          1.503994\n",
      "643837     0.060807          1.467667\n",
      "101928626  0.054134          1.306600\n",
      "107984847  0.053432          1.289662\n",
      "84069      0.053022          1.279757\n",
      "84808      0.052472          1.266478\n",
      "112268260  0.052468          1.266381\n",
      "339451     0.052222          1.260447\n",
      "\n",
      "Relevant genes with contributions for PC9:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "102466751  0.481106         11.533555\n",
      "84808      0.421750         10.110606\n",
      "79501      0.359342          8.614494\n",
      "729759     0.303013          7.264134\n",
      "102725121  0.257095          6.163341\n",
      "645520     0.253049          6.066349\n",
      "26155      0.203035          4.867358\n",
      "79854      0.185500          4.446991\n",
      "100302278  0.160603          3.850142\n",
      "105378947  0.151370          3.628786\n",
      "729737     0.142976          3.427564\n",
      "100287102  0.138066          3.309855\n",
      "102723897  0.131453          3.151333\n",
      "101928626  0.121755          2.918827\n",
      "100288069  0.088476          2.121033\n",
      "107984850  0.079205          1.898784\n",
      "284593     0.064491          1.546038\n",
      "81399      0.064144          1.537727\n",
      "100996442  0.063592          1.524492\n",
      "112268260  0.058844          1.410663\n",
      "\n",
      "Relevant genes with contributions for PC10:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "148398     0.734504         20.411603\n",
      "26155      0.337777          9.386685\n",
      "102723897  0.284212          7.898144\n",
      "102725121  0.222556          6.184762\n",
      "84808      0.200066          5.559769\n",
      "79501      0.176944          4.917209\n",
      "79854      0.150083          4.170758\n",
      "107985728  0.149241          4.147358\n",
      "645520     0.144472          4.014811\n",
      "100130417  0.102831          2.857622\n",
      "107985730  0.101318          2.815592\n",
      "84069      0.087178          2.422638\n",
      "729737     0.086518          2.404315\n",
      "102466751  0.083440          2.318755\n",
      "107984850  0.083351          2.316286\n",
      "105378947  0.065922          1.831943\n",
      "107984847  0.065147          1.810415\n",
      "100302278  0.062824          1.745868\n",
      "400728     0.053478          1.486123\n",
      "\n",
      "Relevant genes with contributions for PC11:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "102723897  0.831735         25.420691\n",
      "148398     0.299124          9.142263\n",
      "105378947  0.212142          6.483775\n",
      "84808      0.186033          5.685796\n",
      "79501      0.160629          4.909387\n",
      "100287102  0.148342          4.533842\n",
      "107984849  0.123526          3.775372\n",
      "100288069  0.105950          3.238188\n",
      "653635     0.091654          2.801264\n",
      "102725121  0.079129          2.418445\n",
      "100302278  0.077365          2.364526\n",
      "100996442  0.070200          2.145551\n",
      "284600     0.069068          2.110970\n",
      "107985728  0.068207          2.084626\n",
      "101928626  0.063787          1.949535\n",
      "400728     0.055846          1.706834\n",
      "79854      0.052024          1.590033\n",
      "102465909  0.049159          1.502460\n",
      "339451     0.048144          1.471442\n",
      "107984850  0.043956          1.343449\n",
      "84069      0.042965          1.313146\n",
      "112268260  0.041768          1.276583\n",
      "\n",
      "Relevant genes with contributions for PC12:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "79501      0.548882         13.532288\n",
      "107985730  0.517676         12.762939\n",
      "79854      0.351494          8.665841\n",
      "100302278  0.215770          5.319661\n",
      "645520     0.201871          4.976990\n",
      "81399      0.181855          4.483496\n",
      "102466751  0.168031          4.142678\n",
      "105378947  0.150330          3.706284\n",
      "102723897  0.131788          3.249127\n",
      "26155      0.120833          2.979040\n",
      "100288069  0.102320          2.522630\n",
      "100287102  0.099116          2.443644\n",
      "84069      0.092378          2.277523\n",
      "729759     0.091116          2.246390\n",
      "148398     0.089295          2.201515\n",
      "107984849  0.083173          2.050559\n",
      "107984847  0.079840          1.968393\n",
      "284600     0.077117          1.901257\n",
      "400728     0.076358          1.882563\n",
      "113219467  0.071466          1.761937\n",
      "729737     0.068401          1.686376\n",
      "100132287  0.064599          1.592645\n",
      "102725121  0.060367          1.488305\n",
      "\n",
      "Relevant genes with contributions for PC13:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "81399      0.452168          9.418009\n",
      "79501      0.359148          7.480539\n",
      "107985730  0.301935          6.288869\n",
      "105378947  0.288008          5.998794\n",
      "84069      0.263765          5.493840\n",
      "84808      0.218179          4.544348\n",
      "339451     0.197642          4.116600\n",
      "107984849  0.192366          4.006710\n",
      "284600     0.184937          3.851974\n",
      "100287102  0.178516          3.718222\n",
      "101928626  0.178052          3.708569\n",
      "102725121  0.145491          3.030370\n",
      "100288069  0.142972          2.977906\n",
      "102466751  0.133268          2.775789\n",
      "653635     0.127367          2.652882\n",
      "26155      0.122863          2.559069\n",
      "107984850  0.121252          2.525509\n",
      "643837     0.117006          2.437075\n",
      "113219467  0.116587          2.428348\n",
      "102723897  0.105345          2.194184\n",
      "100996442  0.096777          2.015732\n",
      "100133331  0.095700          1.993288\n",
      "100302278  0.081781          1.703371\n",
      "400728     0.076794          1.599503\n",
      "\n",
      "Relevant genes with contributions for PC14:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "81399      0.695324         19.647121\n",
      "79501      0.351115          9.921120\n",
      "79854      0.344501          9.734241\n",
      "107984849  0.269574          7.617096\n",
      "729737     0.230982          6.526641\n",
      "101928626  0.187838          5.307549\n",
      "100302278  0.159611          4.509970\n",
      "107985730  0.116265          3.285187\n",
      "100132287  0.110311          3.116954\n",
      "148398     0.105224          2.973227\n",
      "339451     0.096268          2.720154\n",
      "100287102  0.081090          2.291283\n",
      "284600     0.062366          1.762217\n",
      "107984841  0.060118          1.698706\n",
      "729759     0.058040          1.639995\n",
      "100996442  0.056565          1.598309\n",
      "105378580  0.055593          1.570838\n",
      "400728     0.053194          1.503043\n",
      "112268260  0.048005          1.356436\n",
      "\n",
      "Relevant genes with contributions for PC15:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "645520     0.832081         28.091242\n",
      "107985730  0.306634         10.352038\n",
      "100302278  0.243566          8.222833\n",
      "107984849  0.228365          7.709640\n",
      "79501      0.188752          6.372298\n",
      "79854      0.127332          4.298758\n",
      "148398     0.098049          3.310145\n",
      "84069      0.071607          2.417454\n",
      "729737     0.066413          2.242117\n",
      "100132287  0.065573          2.213764\n",
      "102725121  0.055273          1.866034\n",
      "113219467  0.049961          1.686707\n",
      "339451     0.049687          1.677458\n",
      "102723897  0.048823          1.648280\n",
      "84808      0.046078          1.555592\n",
      "107985728  0.041979          1.417224\n",
      "100996442  0.041541          1.402418\n",
      "100287102  0.038836          1.311123\n",
      "653635     0.036270          1.224494\n",
      "\n",
      "Relevant genes with contributions for PC16:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "107984849  0.604363         15.969003\n",
      "84069      0.513639         13.571800\n",
      "84808      0.309420          8.175763\n",
      "339451     0.247943          6.551359\n",
      "148398     0.229771          6.071210\n",
      "79501      0.158484          4.187590\n",
      "107985730  0.144831          3.826853\n",
      "645520     0.130760          3.455045\n",
      "102723897  0.123405          3.260706\n",
      "81399      0.109115          2.883120\n",
      "400728     0.108380          2.863706\n",
      "79854      0.087858          2.321446\n",
      "105378947  0.084968          2.245089\n",
      "101928626  0.080522          2.127613\n",
      "107984850  0.069837          1.845280\n",
      "100132287  0.069265          1.830189\n",
      "100288069  0.059523          1.572766\n",
      "102725121  0.057251          1.512735\n",
      "100133331  0.047625          1.258382\n",
      "100302278  0.047541          1.256163\n",
      "107984841  0.046118          1.218557\n",
      "653635     0.046076          1.217462\n",
      "\n",
      "Relevant genes with contributions for PC17:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "100302278  0.635921         19.662424\n",
      "101928626  0.609497         18.845388\n",
      "107985730  0.354840         10.971523\n",
      "284600     0.120079          3.712782\n",
      "729737     0.108796          3.363927\n",
      "100132287  0.094571          2.924093\n",
      "84069      0.087549          2.706992\n",
      "100287102  0.077571          2.398466\n",
      "107984849  0.067820          2.096970\n",
      "105378947  0.067652          2.091777\n",
      "148398     0.061903          1.914010\n",
      "653635     0.059203          1.830524\n",
      "105378580  0.050478          1.560768\n",
      "643837     0.048942          1.513277\n",
      "107984850  0.046775          1.446250\n",
      "400728     0.046114          1.425838\n",
      "100130417  0.044516          1.376417\n",
      "107984847  0.044297          1.369639\n",
      "102466751  0.042311          1.308247\n",
      "100133331  0.039540          1.222574\n",
      "81399      0.038891          1.202484\n",
      "79501      0.038295          1.184058\n",
      "102465909  0.037219          1.150796\n",
      "100996442  0.036950          1.142470\n",
      "107984841  0.034576          1.069075\n",
      "\n",
      "Relevant genes with contributions for PC18:\n",
      "            Loading  Contribution (%)\n",
      "GeneID                               \n",
      "107984849  0.500616         13.737981\n",
      "101928626  0.482272         13.234596\n",
      "100302278  0.464029         12.733965\n",
      "84069      0.290728          7.978204\n",
      "100288069  0.213076          5.847272\n",
      "107985730  0.186932          5.129812\n",
      "81399      0.158327          4.344831\n",
      "79854      0.129121          3.543372\n",
      "729737     0.126932          3.483300\n",
      "339451     0.124861          3.426469\n",
      "645520     0.116194          3.188611\n",
      "100130417  0.105965          2.907913\n",
      "84808      0.102188          2.804271\n",
      "113219467  0.070204          1.926538\n",
      "112268260  0.065813          1.806060\n",
      "284600     0.056094          1.539351\n",
      "400728     0.051650          1.417397\n"
     ]
    }
   ],
   "source": [
    "expression_cols = merged_data.columns[1:39]\n",
    "expression_data = merged_data[expression_cols].apply(pd.to_numeric, errors='coerce')\n",
    "\n",
    "pca_target_variance = 0.75\n",
    "pca_full = PCA()\n",
    "expression_data_filled = expression_data.apply(lambda row: row.bfill().ffill(), axis=1)\n",
    "nan_counts = expression_data.isna().sum().sum()\n",
    "nan_counts\n",
    "\n",
    "pca_full.fit(expression_data)\n",
    "\n",
    "# Determine number of components needed to explain 75% variance\n",
    "cumulative_variance = pca_full.explained_variance_ratio_.cumsum()\n",
    "n_components_75 = np.argmax(cumulative_variance >= pca_target_variance) + 1\n",
    "\n",
    "# Apply PCA with the determined number of components\n",
    "pca_final = PCA(n_components=n_components_75)\n",
    "pca_result_final = pca_final.fit_transform(expression_data)\n",
    "\n",
    "# Retrieve top contributing genes for each principal component\n",
    "components_df = pd.DataFrame(\n",
    "    pca_final.components_,\n",
    "    columns=merged_data['GeneID'][:expression_data.shape[1]],\n",
    "    index=[f\"PC{i+1}\" for i in range(n_components_75)]\n",
    ")\n",
    "\n",
    "def get_relevant_genes_with_contribution(pc, cumulative_threshold=0.9):\n",
    "    # Absolute loadings for the specified principal component\n",
    "    loadings = components_df.iloc[pc-1].abs().sort_values(ascending=False)\n",
    "\n",
    "    # Calculate cumulative variance contribution by each gene\n",
    "    cumulative_variance = loadings.cumsum() / loadings.sum()\n",
    "\n",
    "    # Select genes up to the cumulative threshold\n",
    "    relevant_genes = loadings[cumulative_variance <= cumulative_threshold]\n",
    "\n",
    "    # Calculate the contribution percentage for each gene\n",
    "    contribution_percentages = (relevant_genes / loadings.sum()) * 100\n",
    "\n",
    "    return pd.DataFrame({\n",
    "        'Loading': relevant_genes,\n",
    "        'Contribution (%)': contribution_percentages\n",
    "    })\n",
    "\n",
    "# Example usage\n",
    "for pc in range(1, n_components_75 + 1):\n",
    "    print(f\"\\nRelevant genes with contributions for PC{pc}:\")\n",
    "    print(get_relevant_genes_with_contribution(pc))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "PC1 - Top Keywords:\n",
      " ['protein', 'coding', 'intergenic', 'long', 'non', 'rna', '6859', 'microrna', '100288069', 'by']\n",
      "PC1 - Top Topic Words:\n",
      " ['replaced', '100288069', 'microrna', '6859', 'intergenic', 'rna', 'long', 'coding', 'non', 'protein']\n",
      "\n",
      "PC2 - Top Keywords:\n",
      " ['family', 'microrna', 'member', 'pseudogene', 'like', 'protein', 'olfactory', 'receptor', 'subfamily', '11']\n",
      "PC2 - Top Topic Words:\n",
      " ['box', 'receptor', 'subfamily', 'olfactory', 'protein', 'like', 'pseudogene', 'member', 'microrna', 'family']\n",
      "\n",
      "PC3 - Top Keywords:\n",
      " ['like', 'alpha', 'chain', 'collagen', 'iii', 'associated', 'epr1', 'extensin', 'noc2', 'nucleolar']\n",
      "PC3 - Top Topic Words:\n",
      " ['transcriptional', 'extensin', 'epr1', 'associated', 'noc2', 'iii', 'collagen', 'chain', 'alpha', 'like']\n",
      "\n",
      "PC4 - Top Keywords:\n",
      " ['like', 'alpha', 'chain', 'collagen', 'iii', 'epr1', 'extensin', 'proline', 'protein', 'rich']\n",
      "PC4 - Top Topic Words:\n",
      " ['epr1', 'extensin', 'proline', 'protein', 'rich', 'alpha', 'chain', 'collagen', 'iii', 'like']\n",
      "\n",
      "PC5 - Top Keywords:\n",
      " ['6859', 'microrna', 'alpha', 'chain', 'collagen', 'iii', 'like', 'and', 'esrr', 'induced']\n",
      "PC5 - Top Topic Words:\n",
      " ['muscle', 'ppargc1', 'regulator', 'alpha', 'chain', 'collagen', 'iii', 'like', '6859', 'microrna']\n",
      "\n",
      "PC6 - Top Keywords:\n",
      " ['11', '17', 'box', 'dead', 'helicase', 'like', 'pseudogene']\n",
      "PC6 - Top Topic Words:\n",
      " ['11', '17', 'box', 'dead', 'helicase', 'like', 'pseudogene']\n",
      "\n",
      "PC7 - Top Keywords:\n",
      " ['associated', 'like', 'noc2', 'nucleolar', 'repressor', 'transcriptional']\n",
      "PC7 - Top Topic Words:\n",
      " ['associated', 'like', 'noc2', 'nucleolar', 'repressor', 'transcriptional']\n",
      "\n",
      "PC8 - Top Keywords:\n",
      " ['gene', 'host', 'mir1302', '29', 'family', 'member', 'olfactory', 'receptor', 'subfamily', '115']\n",
      "PC8 - Top Topic Words:\n",
      " ['rna', '29', 'family', 'member', 'olfactory', 'receptor', 'subfamily', 'gene', 'host', 'mir1302']\n",
      "\n",
      "PC9 - Top Keywords:\n",
      " ['family', 'member', 'olfactory', 'receptor', 'subfamily', '6859', 'microrna', '29', '138', 'sequence']\n",
      "PC9 - Top Topic Words:\n",
      " ['138', 'with', '29', 'microrna', '6859', 'receptor', 'subfamily', 'olfactory', 'member', 'family']\n",
      "\n",
      "PC10 - Top Keywords:\n",
      " ['11', 'alpha', 'containing', 'domain', 'motif', 'sterile']\n",
      "PC10 - Top Topic Words:\n",
      " ['11', 'alpha', 'containing', 'domain', 'motif', 'sterile']\n",
      "\n",
      "PC11 - Top Keywords:\n",
      " ['family', 'homolog', 'protein', 'pseudogene', 'was']\n",
      "PC11 - Top Topic Words:\n",
      " ['family', 'homolog', 'protein', 'pseudogene', 'was']\n",
      "\n",
      "PC12 - Top Keywords:\n",
      " ['gene', 'host', 'mir1302', 'family', 'member', 'olfactory', 'receptor', 'subfamily', '115', 'coding']\n",
      "PC12 - Top Topic Words:\n",
      " ['protein', 'rna', 'family', 'member', 'olfactory', 'receptor', 'subfamily', 'gene', 'host', 'mir1302']\n",
      "\n",
      "PC13 - Top Keywords:\n",
      " ['family', 'member', 'olfactory', 'receptor', 'subfamily', 'gene', 'host', 'mir1302', '16', 'containing']\n",
      "PC13 - Top Topic Words:\n",
      " ['n1', '16', 'host', 'gene', 'mir1302', 'member', 'olfactory', 'family', 'receptor', 'subfamily']\n",
      "\n",
      "PC14 - Top Keywords:\n",
      " ['family', 'member', 'olfactory', 'receptor', 'subfamily', '16']\n",
      "PC14 - Top Topic Words:\n",
      " ['16', 'family', 'olfactory', 'receptor', 'member', 'subfamily']\n",
      "\n",
      "PC15 - Top Keywords:\n",
      " ['138', 'family', 'member', 'sequence', 'similarity', 'with']\n",
      "PC15 - Top Topic Words:\n",
      " ['138', 'family', 'member', 'sequence', 'similarity', 'with']\n",
      "\n",
      "PC16 - Top Keywords:\n",
      " ['100288069', 'by', 'id', 'replaced', 'containing', 'domain', 'homology', 'n1', 'pleckstrin', 'and']\n",
      "PC16 - Top Topic Words:\n",
      " ['regulator', 'containing', 'domain', 'homology', 'n1', 'pleckstrin', '100288069', 'by', 'id', 'replaced']\n",
      "\n",
      "PC17 - Top Keywords:\n",
      " ['1302', 'loc101928626', 'microrna', 'uncharacterized', 'gene', 'host', 'mir1302']\n",
      "PC17 - Top Topic Words:\n",
      " ['gene', 'host', 'mir1302', '1302', 'loc101928626', 'microrna', 'uncharacterized']\n",
      "\n",
      "PC18 - Top Keywords:\n",
      " ['1302', 'loc101928626', 'microrna', 'uncharacterized', '100288069', 'by', 'id', 'replaced', 'containing', 'domain']\n",
      "PC18 - Top Topic Words:\n",
      " ['n1', 'pleckstrin', '100288069', 'by', 'id', 'replaced', '1302', 'loc101928626', 'microrna', 'uncharacterized']\n"
     ]
    }
   ],
   "source": [
    "from sklearn.feature_extraction.text import TfidfVectorizer\n",
    "from sklearn.decomposition import LatentDirichletAllocation as LDA\n",
    "from bertopic import BERTopic\n",
    "import pandas as pd\n",
    "\n",
    "# Ensure GeneID is an integer for merging purposes\n",
    "merged_data['GeneID'] = merged_data['GeneID'].astype(int)\n",
    "\n",
    "# Step 1: Fetch descriptions for top genes in each PCA component\n",
    "top_descriptions = {}\n",
    "for pc, genes in relevant_genes.items():\n",
    "    descriptions = merged_data[merged_data['GeneID'].isin(genes.index)]['Description'].dropna().tolist()\n",
    "    top_descriptions[pc] = descriptions\n",
    "\n",
    "# Step 2: Apply TF-IDF to extract keywords from descriptions\n",
    "tfidf_vectorizer = TfidfVectorizer(max_features=100)\n",
    "pc_keywords = {}\n",
    "\n",
    "for pc, descriptions in top_descriptions.items():\n",
    "    tfidf_matrix = tfidf_vectorizer.fit_transform(descriptions)\n",
    "    feature_names = tfidf_vectorizer.get_feature_names_out()\n",
    "\n",
    "    # Sum up TF-IDF scores for each term across all descriptions in this component\n",
    "    scores = tfidf_matrix.sum(axis=0).A1\n",
    "    keywords = sorted(zip(feature_names, scores), key=lambda x: x[1], reverse=True)[:10]  # Top 10 keywords\n",
    "    pc_keywords[pc] = [keyword for keyword, score in keywords]\n",
    "\n",
    "# Step 3: Topic Modeling with LDA for theme-based interpretation\n",
    "lda_model = LDA(n_components=1)  # 1 topic per component\n",
    "\n",
    "pc_topics = {}\n",
    "for pc, descriptions in top_descriptions.items():\n",
    "    tfidf_matrix = tfidf_vectorizer.fit_transform(descriptions)\n",
    "    lda_model.fit(tfidf_matrix)\n",
    "\n",
    "    # Get the top words for the single topic in this PCA component\n",
    "    topic_words = [tfidf_vectorizer.get_feature_names_out()[i] for i in lda_model.components_[0].argsort()[-10:]]\n",
    "    pc_topics[pc] = topic_words\n",
    "\n",
    "# Step 4: Clustering with BERTopic for descriptive cluster titles\n",
    "bertopic_model = BERTopic()\n",
    "all_descriptions = [desc for descs in top_descriptions.values() for desc in descs]\n",
    "topic_model, probabilities = bertopic_model.fit_transform(all_descriptions)\n",
    "\n",
    "# Generate cluster titles for each topic in BERTopic model\n",
    "cluster_titles = bertopic_model.get_topic_info().set_index(\"Topic\")[\"Name\"].to_dict()\n",
    "\n",
    "# Display summarized keywords, topic words, and cluster titles for each component\n",
    "for pc in pc_keywords:\n",
    "    print(f\"\\n{pc} - Top Keywords:\\n\", pc_keywords[pc])\n",
    "    print(f\"{pc} - Top Topic Words:\\n\", pc_topics[pc])\n",
    "    print(f\"{pc} - Cluster Title:\\n\", cluster_titles.get(pc, \"No Title\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
